Adding attribute by team name

nathanaelhoun opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

- [x] Add attributes by team membership
- [x] Customise settings from UI with a team picker
- [x] Update README.md

I didn't tested if it doesn't break the GroupID version, because I don't have access to an E20 server, but I tested the others cases.

Inspired by #15 

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
No ticket

hanzei commented:
@aaronrothschild I would like to hear you thoughts on this feature proposal.

marianunez commented:
@nathanaelhoun Thank you for your contribution! Just to confirm, this new feature will allow custom attributes to be added to all user that belongs to the configured team(s)?

> Update README.md (I don't know how to update the second screenshot here)

To update the image, you can either update it directly in Github and upload the image that you would want (you can drag and drop) or you can create a `docs` folder to add your image and then reference it as a relative path in url markdown added in the Readme - you can see an example of that [here](https://github.com/mattermost/mattermost-plugin-demo/blob/master/webapp/README.md#user-attributes)

nathanaelhoun commented:
Thanks for the tip @marianunez !

Yes exactly: every member of the team will have the team attribute. We use it because we don't have the AD groups in our Mattermost instance.

aaronrothschild commented:
How does this work when you belong to 20+ teams @nathanaelhoun ?  Could you post some screenshots here?

nathanaelhoun commented:
@aaronrothschild This looks like this with one team for each letter in the alphabet. Looks pretty messy I guess.
![image](https://user-images.githubusercontent.com/45119518/84799906-3f313e80-affd-11ea-8ff9-840a58971abf.png)

A little clarification: the custom attribute only shows when it's configured for the specified team (it is not automatic for all the teams)
![image](https://user-images.githubusercontent.com/45119518/84798212-f7a9b300-affa-11ea-8361-19e2a5b318e8.png)

nathanaelhoun commented:
Thanks for the feeback @marianunez! It took me some time (React is totally new to me, I had to understand a few things), and I force-pushed as I rewrote everything I had changed previously. 
Now there is a team selector in the UI, and teams are compared with their ids.
![image](https://user-images.githubusercontent.com/45119518/85234745-ef44e400-b40f-11ea-836e-6ecac8adbd77.png)

marianunez commented on webapp/src/components/admin_settings/teams_input/teams_input.jsx:
```diff
@@ -0,0 +1,93 @@
+// TeamsInput searches and selects teams displayed by display_name.
+// Teams prop can handle the team object or strings directly if the team object is not available.
+// Returns the selected team ids in the `OnChange` value parameter.
+export default class TeamsInput extends React.Component {
```
Not in scope with this PR, but it would be interesting to see if we can generalize this input component. Beside `formatOptionLabel` and `getOptionValue`, everything else is identical. It would be great if we would pass those as props and it can be reused for both teams and users. It could also possibly solve https://github.com/mattermost/mattermost-plugin-custom-attributes/issues/28

marianunez replied:
@nathanaelhoun would you be interested in giving https://github.com/mattermost/mattermost-plugin-custom-attributes/issues/28 a shot and possibly exploring the approach above that would apply for users, groups and teams?

mickmister commented on webapp/src/components/admin_settings/teams_input/teams_input.jsx:
```diff
@@ -0,0 +1,93 @@
+        });
+    }, 150);
+
+    teamsLoader = (term, callback) => {
+        try {
+            this.debouncedSearchTeams(term, callback);
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(error);
+            callback([]);
+        }
+    };
```
I think we can get rid of this function and just pass in the debounced version to the child component. The `debouncedSearchTeams` won't throw an error for an API response error anyway, so this catch block won't run.

nathanaelhoun replied:
:+1:

nathanaelhoun commented on webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx:
```diff
@@ -58,6 +62,27 @@ export default class CustomAttribute extends React.Component {
+            return;
+        }
+
+        var teamsData = [];
+        for (let i = 0; i < this.props.teams.length; i++) {
+            teamsData.push(this.props.actions.getTeam(this.props.teams[i]));
+        }
```
Yes !

nathanaelhoun commented on webapp/src/components/admin_settings/custom_attribute/custom_attribute.jsx:
```diff
@@ -58,6 +62,27 @@ export default class CustomAttribute extends React.Component {
+        }
+
+        var teams = [];
+        await Promise.all(teamsData).then((values) => {
+            values.forEach((value) => {
+                teams.push(value.data);
+            });
+        });
```
:+1:

nathanaelhoun commented on webapp/src/components/admin_settings/teams_input/teams_input.jsx:
```diff
@@ -0,0 +1,93 @@
+// TeamsInput searches and selects teams displayed by display_name.
+// Teams prop can handle the team object or strings directly if the team object is not available.
+// Returns the selected team ids in the `OnChange` value parameter.
+export default class TeamsInput extends React.Component {
```
:+1:

nathanaelhoun commented on webapp/src/components/admin_settings/teams_input/teams_input.jsx:
```diff
@@ -0,0 +1,93 @@
+        return option;
+    }
+
+    debouncedSearchTeams = debounce((term, callback) => {
+        this.props.actions.searchTeams(term).then(({data}) => {
+            callback(data);
+        }).catch(() => {
+            // eslint-disable-next-line no-console
+            console.error('Error searching teams in custom attribute settings dropdown.');
+            callback([]);
+        });
+    }, 150);
```
Good catch !

nathanaelhoun commented:
Thanks for the review @mickmister :smile:

mickmister commented on webapp/src/components/admin_settings/teams_input/teams_input.jsx:
```diff
@@ -0,0 +1,86 @@
+        this.props.actions.searchTeams(term).then(({data, error}) => {
+            if (error) {
+                // eslint-disable-next-line no-console
+                console.error('Error searching team in custom attribute settings dropdown.' + error.message);
```
Non-blocking request to add a space after the first sentence
```suggestion
                console.error('Error searching team in custom attribute settings dropdown. ' + error.message);
```

nathanaelhoun replied:
This is what happens when I don't use a code formatter :sweat_smile:

marianunez commented:
@aaronrothschild any concerns with moving forward with this feature?

nathanaelhoun commented:
Hi! Is there any news on this?

marianunez commented:
@nathanaelhoun apologies on the delay on this. 

@aaronrothschild gentle ping to `PM Review` so we can go ahead with `QA Review`.

mickmister commented:
>  Looks good, I am worried about too many Teams when there is a bunch setup, but that can be conquered later.

@aaronrothschild Do you think this needs its own ticket?

Thanks so much @nathanaelhoun!
